File: sdks/go/provider_test.go
package flagcel

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/open-feature/go-sdk/openfeature"
	"github.com/picunada/flagcel/evalcore"
)

func TestProviderFetchesDefinitionsAndEvaluates(t *testing.T) {
	requests := 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requests++
		if got := r.Header.Get("Authorization"); got != "Bearer secret" {
			t.Fatalf("Authorization = %q, want Bearer secret", got)
		}
		w.Header().Set("ETag", `"v1"`)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"message": "success",
			"data": definitionsFixture("checkout-copy", evalcore.ValueTypeString, "control", []evalcore.Rule{
				{
					ID:         "pro",
					Expression: `user.tier == "pro"`,
					Rollout:    evalcore.Rollout{Percentage: 100},
					Value:      "pro-copy",
				},
			}),
		})
	}))
	defer server.Close()

	provider, err := NewProvider(server.URL, "secret", WithPollInterval(time.Hour))
	if err != nil {
		t.Fatalf("NewProvider() error = %v", err)
	}
	if err := provider.Init(openfeature.EvaluationContext{}); err != nil {
		t.Fatalf("Init() error = %v", err)
	}
	defer provider.Shutdown()

	detail := provider.StringEvaluation(context.Background(), "checkout-copy", "fallback", openfeature.FlattenedContext{
		"user": map[string]any{"tier": "pro"},
	})
	if detail.Value != "pro-copy" {
		t.Fatalf("Value = %q, want pro-copy", detail.Value)
	}
	if detail.Reason != openfeature.TargetingMatchReason {
		t.Fatalf("Reason = %q, want %q", detail.Reason, openfeature.TargetingMatchReason)
	}
	if detail.Variant != "pro" {
		t.Fatalf("Variant = %q, want pro", detail.Variant)
	}
	if requests != 1 {
		t.Fatalf("requests = %d, want 1", requests)
	}
}

func TestProviderHonorsETagAndKeepsLastKnownDefinitions(t *testing.T) {
	requests := 0
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		requests++
		switch requests {
		case 1:
			w.Header().Set("ETag", `"v1"`)
			_ = json.NewEncoder(w).Encode(map[string]any{
				"message": "success",
				"data":    definitionsFixture("enabled", evalcore.ValueTypeBoolean, true, nil),
			})
		case 2:
			if got := r.Header.Get("If-None-Match"); got != `"v1"` {
				t.Fatalf("If-None-Match = %q, want %q", got, `"v1"`)
			}
			w.Header().Set("ETag", `"v1"`)
			w.WriteHeader(http.StatusNotModified)
		default:
			http.Error(w, "temporarily unavailable", http.StatusServiceUnavailable)
		}
	}))
	defer server.Close()

	provider, err := NewProvider(server.URL, "", WithPollInterval(time.Hour))
	if err != nil {
		t.Fatalf("NewProvider() error = %v", err)
	}
	if err := provider.Init(openfeature.EvaluationContext{}); err != nil {
		t.Fatalf("Init() error = %v", err)
	}
	defer provider.Shutdown()

	if err := provider.refresh(context.Background()); err != nil {
		t.Fatalf("refresh() 304 error = %v", err)
	}
	if err := provider.refresh(context.Background()); err == nil {
		t.Fatal("refresh() error = nil, want fetch error")
	}

	detail := provider.BooleanEvaluation(context.Background(), "enabled", false, openfeature.FlattenedContext{})
	if !detail.Value {
		t.Fatal("Value = false, want last-known true")
	}
	if detail.ProviderResolutionDetail.Error() != nil {
		t.Fatalf("ResolutionError = %v, want empty for successful last-known evaluation", detail.ProviderResolutionDetail.Error())
	}
}

func TestProviderFallsBackToDefaultWhenInitialFetchFails(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "no definitions", http.StatusServiceUnavailable)
	}))
	defer server.Close()

	provider, err := NewProvider(server.URL, "", WithPollInterval(time.Hour))
	if err != nil {
		t.Fatalf("NewProvider() error = %v", err)
	}
	if err := provider.Init(openfeature.EvaluationContext{}); err != nil {
		t.Fatalf("Init() error = %v", err)
	}
	defer provider.Shutdown()

	detail := provider.BooleanEvaluation(context.Background(), "missing", true, openfeature.FlattenedContext{})
	if !detail.Value {
		t.Fatal("Value = false, want default true")
	}
	if detail.ResolutionError.Error() == "" {
		t.Fatal("ResolutionError is empty, want provider not ready error")
	}
}

func TestProviderTypeMismatchFallsBackToDefault(t *testing.T) {
	provider := &Provider{}
	if err := provider.cache.store(definitionsFixture("title", evalcore.ValueTypeString, "hello", nil), `"v1"`); err != nil {
		t.Fatalf("store() error = %v", err)
	}

	detail := provider.BooleanEvaluation(context.Background(), "title", true, openfeature.FlattenedContext{})
	if !detail.Value {
		t.Fatal("Value = false, want default true")
	}
	if detail.ResolutionError.Error() == "" {
		t.Fatal("ResolutionError is empty, want type mismatch")
	}
}

func definitionsFixture(key string, valueType evalcore.ValueType, defaultValue any, rules []evalcore.Rule) evalcore.Definitions {
	return evalcore.Definitions{
		Flags: []evalcore.FlagDefinition{
			{
				FlagConfig: evalcore.FlagConfig{
					Key:          key,
					Type:         valueType,
					Enabled:      true,
					DefaultValue: defaultValue,
					Rules:        rules,
				},
				ContextSchema: &evalcore.ContextSchema{
					Fields: []evalcore.ContextField{
						{Path: "user.tier", Type: evalcore.ContextTypeString},
					},
				},
			},
		},
	}
}

File: sdks/js/src/provider.test.ts
import { ErrorCode, StandardResolutionReasons, type EvaluationContext } from "@openfeature/server-sdk";
import { describe, expect, it, vi } from "vitest";

import { FlagcelProvider } from "./provider.js";
import type { EvaluationResult } from "./types.js";

const logger = {
  error: vi.fn(),
  warn: vi.fn(),
  info: vi.fn(),
  debug: vi.fn(),
};

describe("FlagcelProvider", () => {
  it("sends authenticated server-side evaluation requests", async () => {
    const fetchMock = vi.fn(async (input: URL | RequestInfo, init?: RequestInit) => {
      expect(String(input)).toBe("https://flagcel.test/api/v1/eval/enabled");
      const headers = new Headers(init?.headers);
      expect(init?.method).toBe("POST");
      expect(headers.get("Authorization")).toBe("Bearer secret");
      expect(headers.get("Content-Type")).toBe("application/json");
      expect(JSON.parse(String(init?.body))).toEqual({
        context: { targetingKey: "user-123" },
      });
      return jsonResponse({
        key: "enabled",
        value: true,
        value_type: "boolean",
        reason: "matched_rule",
        variant: "targeted",
      });
    });

    const provider = new FlagcelProvider({
      endpoint: "https://flagcel.test/api/v1",
      apiKey: "secret",
      fetch: fetchMock,
    });
    await provider.initialize();

    const detail = await provider.resolveBooleanEvaluation(
      "enabled",
      false,
      { targetingKey: "user-123" },
      logger,
    );

    expect(detail).toMatchObject({
      value: true,
      reason: StandardResolutionReasons.TARGETING_MATCH,
      variant: "targeted",
    });
    expect(fetchMock).toHaveBeenCalledTimes(1);
  });

  it("resolves string, number, and object values from server responses", async () => {
    const provider = new FlagcelProvider({
      endpoint: "https://flagcel.test/api/v1",
      fetch: async (input: URL | RequestInfo) => {
        const key = String(input).split("/").pop();
        if (key === "title") {
          return jsonResponse({ key, value: "hello", value_type: "string", reason: "default_no_match" });
        }
        if (key === "count") {
          return jsonResponse({ key, value: 42, value_type: "number", reason: "default_no_match" });
        }
        return jsonResponse({ key, value: { plan: "pro" }, value_type: "json", reason: "default_no_match" });
      },
    });

    await expect(provider.resolveStringEvaluation("title", "", {}, logger)).resolves.toMatchObject({
      value: "hello",
      reason: StandardResolutionReasons.DEFAULT,
    });
    await expect(provider.resolveNumberEvaluation("count", 0, {}, logger)).resolves.toMatchObject({
      value: 42,
      reason: StandardResolutionReasons.DEFAULT,
    });
    await expect(provider.resolveObjectEvaluation("payload", {}, {}, logger)).resolves.toMatchObject({
      value: { plan: "pro" },
      reason: StandardResolutionReasons.DEFAULT,
    });
  });

  it("returns default on type mismatch", async () => {
    const provider = initializedProvider({ key: "title", value: "hello", value_type: "string" });

    const detail = await provider.resolveBooleanEvaluation("title", true, {}, logger);

    expect(detail).toMatchObject({
      value: true,
      reason: StandardResolutionReasons.ERROR,
      errorCode: ErrorCode.TYPE_MISMATCH,
    });
  });

  it("returns default on not found responses", async () => {
    const provider = new FlagcelProvider({
      endpoint: "https://flagcel.test/api/v1",
      fetch: async () =>
        new Response(JSON.stringify({ error: { code: "FLAG_NOT_FOUND", message: "Flag not found" } }), {
          status: 404,
          headers: { "Content-Type": "application/json" },
        }),
    });

    const detail = await provider.resolveBooleanEvaluation("missing", true, {}, logger);

    expect(detail).toMatchObject({
      value: true,
      reason: StandardResolutionReasons.ERROR,
      errorCode: ErrorCode.FLAG_NOT_FOUND,
      errorMessage: "Flag not found",
    });
  });

  it("returns default on network failures", async () => {
    const provider = new FlagcelProvider({
      endpoint: "https://flagcel.test/api/v1",
      fetch: async () => {
        throw new Error("connection refused");
      },
    });

    const detail = await provider.resolveBooleanEvaluation("enabled", false, {}, logger);

    expect(detail).toMatchObject({
      value: false,
      reason: StandardResolutionReasons.ERROR,
      errorCode: ErrorCode.GENERAL,
      errorMessage: "connection refused",
    });
  });
});

function initializedProvider(result: EvaluationResult): FlagcelProvider {
  return new FlagcelProvider({
    endpoint: "https://flagcel.test/api/v1",
    fetch: async () => jsonResponse(result),
  });
}

function jsonResponse(result: EvaluationResult): Response {
  return new Response(JSON.stringify({ message: "success", data: result }), {
    status: 200,
    headers: { "Content-Type": "application/json" },
  });
}
